src/display.rs

```diff
@@ -538,6 +538,51 @@ where
     f.pad_integral(true, "0x", encoded)
 }
 
+/// Given a `T:` [`fmt::Write`], `HexWriter` implements [`std::io::Write`]
+/// and writes the source bytes to its inner `T` as hex characters.
+#[cfg(any(test, feature = "std"))]
+#[cfg_attr(docsrs, doc(cfg(any(test, feature = "std"))))]
+pub struct HexWriter<T>(T, Case);
+
+#[cfg(any(test, feature = "std"))]
+#[cfg_attr(docsrs, doc(cfg(any(test, feature = "std"))))]
+impl<T> HexWriter<T> {
+    /// Creates a `HexWriter` that writes the source bytes to `dest` as hex characters
+    /// in the given `case`.
+    pub fn new(dest: T, case: Case) -> Self { Self(dest, case) }
+    /// Consumes this `HexWriter` returning the inner `T`.
+    pub fn into_inner(self) -> T { self.0 }
+}
+
+#[cfg(any(test, feature = "std"))]
+#[cfg_attr(docsrs, doc(cfg(any(test, feature = "std"))))]
+impl<T> std::io::Write for HexWriter<T>
+where
+    T: core::fmt::Write,
+{
+    /// # Errors
+    ///
+    /// If no bytes could be written to this `HexWriter`, and the provided buffer is not empty,
+    /// returns [`std::io::ErrorKind::Other`], otherwise returns `Ok`.
+    fn write(&mut self, buf: &[u8]) -> Result<usize, std::io::Error> {
+        let mut n = 0;
+        let table = self.1.table();
+        for byte in buf {
+            let pair = table.byte_to_hex(*byte);
+            if self.0.write_str(&pair).is_err() {
```

Comment:
This is writing two-hex digits (chars), we have no way of knowing if both were written on error or not. So the only option is to write the first one and buffer the second one if there was an error.

Reply:
We could just "buffer" a flag saying to return an error on the next call.

Reply:
> This is writing two-hex digits (chars), we have no way of knowing if both were written on error or not.

I went with `write_str` to tell `T: fmt::Write`: ***either** write 2 chars, **or** return an error, **don't** write only one*. 

Upon error, I assume that zero characters have been written.

I understand this might be a shaky assumption, for `arrayvec::ArrayString` it holds up. Let me know what you think.

>  So the only option is to write the first one and buffer the second one if there was an error.

I am hesitant to use something like `write_char` to write character by character. If the first character write returns ok, but the second character write returns error, my writer has written half a byte, and has no way of telling `std::io::Write` exactly how many bytes it's written because of the unsigned integer return type (as poelstra has pointed out above).

I understand this could also be a problem with `write_str` method currently in use should my assumption about atomic 2 character writes not hold.

> We could just "buffer" a flag saying to return an error immediately on the next call.

Happy to do that yes. Is that the performance issue you've been referring to @Kixunil ?